x-pack/dev-tools/simulate_autoscaling.py

```diff
@@ -0,0 +1,276 @@
+from collections import deque, defaultdict
+import enum
+import heapq
+import math
+import random
+import matplotlib.pyplot as plt
+
+
+START_TIME = 0
+END_TIME = 5*3600
+MEASUREMENT_INTERVAL = 1
+
+AUTO_SCALE_UP_THRESHOLD = 0.9
+AUTO_SCALE_DOWN_THRESHOLD = 0.8
+
+AVG_INFERENCE_TIME = 0.1
+PHYSICAL_CORES_PER_NODE = 4
+
+
+random.seed(170681)
+
+
+class Estimator:
+  def __init__(self, value, variance, smoothing_factor):
+    self.value = value
+    self.variance = variance
+    self.smoothing_factor = smoothing_factor
+
+  def add(self, value, variance, dynamics_changed=False) -> None:
+    process_variance = variance / self.smoothing_factor
+
+    if dynamics_changed or abs(value - self.value)**2 / (self.variance + variance) > 100:
```

Comment:
I would add a comment here for the record...
```suggestion
    # If we know we likely had a change in the quantity we're estimating or the prediction is 10 std off
    # we inject extra noise in the dynamics for this step.
    if dynamics_changed or abs(value - self.value)**2 / (self.variance + variance) > 100:
```

Reply:
One last thought. The estimator will behave badly with outliers: we'll immediately update the state to something in the vicinity of the outlier because of the 10 sigma rule.

For rate this is probably not so likely, but for inference time estimation there could be environmental factors which cause occasional very slow inferences. I would try simulating with occasional 10-20 x inference time values. As a minimum I would probably drop the 10 sigma rule for the inference time Q estimate. The approach which estimates the constant of proportionality between average inference time and allocation count will also be robust to outliers.

Reply:
Good point. This indeed breaks the autoscaler. I've modified the code to address the problem. PTAL.